src/XAD/Jacobian.hpp

```diff
@@ -0,0 +1,138 @@
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#pragma once
+
+#include <XAD/TypeTraits.hpp>
+#include <XAD/XAD.hpp>
+
+#include <functional>
+#include <iterator>
+#include <type_traits>
+#include <vector>
+
+namespace xad
+{
+// adj 2d vector
+template <typename T>
+std::vector<std::vector<T>> computeJacobian(
+    const std::vector<xad::AReal<T>> &vec,
+    std::function<std::vector<xad::AReal<T>>(std::vector<xad::AReal<T>> &)> foo,
+    xad::Tape<T> *tape = xad::Tape<T>::getActive())
+{
+    auto v(vec);
+    std::vector<std::vector<T>> matrix(foo(v).size(), std::vector<T>(v.size(), 0.0));
+    computeJacobian(vec, foo, begin(matrix), end(matrix), tape);
+    return matrix;
+}
+
+// adj iterator
+template <class RowIterator, typename T>
+void computeJacobian(const std::vector<xad::AReal<T>> &vec,
+                     std::function<std::vector<xad::AReal<T>>(std::vector<xad::AReal<T>> &)> foo,
+                     RowIterator first, RowIterator last,
+                     xad::Tape<T> *tape = xad::Tape<T>::getActive())
+{
+    if (std::distance(first->cbegin(), first->cend()) != vec.size())
+        throw OutOfRange("Iterator not allocated enough space (domain)");
+    static_assert(
+        xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
+        "RowIterator must dereference to a type that implements a begin() method");
+    std::unique_ptr<xad::Tape<T>> t;
+    if (!tape)
+    {
+        t = std::unique_ptr<xad::Tape<T>>(new xad::Tape<T>());
+        tape = t.get();
+    }
+
+    auto v(vec);
+
+    tape->registerInputs(v);
+    tape->newRecording();
+    auto y = foo(v);
+    unsigned int domain = static_cast<unsigned int>(vec.size()),
+                 codomain = static_cast<unsigned int>(y.size());
+    if (std::distance(first, last) != codomain)
+        throw OutOfRange("Iterator not allocated enough space (codomain)");
+    tape->registerOutputs(y);
+
+    auto row = first;
+    for (unsigned int i = 0; i < codomain; i++, row++)
+    {
+        auto col = row->begin();
+        derivative(y[i]) = 1.0;
+        tape->computeAdjoints();
+        for (unsigned int j = 0; j < domain; j++, col++) *col = derivative(v[j]);
+        tape->clearDerivatives();
+    }
+}
+
+// fwd 2d vector
+template <typename T>
+std::vector<std::vector<T>> computeJacobian(
+    const std::vector<xad::FReal<T>> &vec,
+    std::function<std::vector<xad::FReal<T>>(std::vector<xad::FReal<T>> &)> foo)
+{
+    auto v(vec);
```

Comment:
Idea, what if we made the 2d vector returning function not call the iterator version? That would reduce the need for excess copies and calls at the expense of some duplicate code, which is honestly a lot easier to solve with helper functions and whatnot.

Reply:
If you just remove `auto v(vec)` and pass `vec` directly to the iterator version, that should be enough.

Reply:
Without copying vec, we can't compute the codomain (since vec is constant and can't be passed to foo) and allocate the space needed for the Jacobian though, which causes seg faults. Note Jacobians aren't always square, so we need both the domain and codomain. We could resize the matrix if needed but then we're suddenly going from copying a vector of size n to n vectors of size n. I really don't see a workaround that doesn't involve either adding a helper function or duplicating some code.

Reply:
understood, let's leave it.